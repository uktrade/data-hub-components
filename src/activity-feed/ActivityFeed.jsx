import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { SPACING } from '@govuk-react/constants'

import Activity from './Activity'
import ActivityFeedHeader from './ActivityFeedHeader'
import ActivityFeedFilters from './ActivityFeedFilters'
import ActivityFeedShowAll from './ActivityFeedShowAll'
import ActivityFeedPagination from './ActivityFeedPagination'

const ActivityFeedContainer = styled('div')`
  margin: ${SPACING.SCALE_2} 0;
`

const ActivityFeedCardList = styled('ol')`
  list-style-type: none;
  padding: 0;
  margin-top: ${SPACING.SCALE_2};

  & > li {
    margin-bottom: ${SPACING.SCALE_2};
  }
`

export default class ActivityFeed extends React.Component {
  static propTypes = {
    onLoadMore: PropTypes.func,
    sendQueryParams: PropTypes.func,
    children: PropTypes.node,
    activities: PropTypes.arrayOf(PropTypes.object),
    activityTypeFilters: PropTypes.object,
    hasMore: PropTypes.bool,
    isLoading: PropTypes.bool,
    actions: PropTypes.node,
    totalActivities: PropTypes.number,
    isGlobalUltimate: PropTypes.bool,
    dnbHierarchyCount: PropTypes.number,
    isTypeFilterFlagEnabled: PropTypes.bool,
    isGlobalUltimateFlagEnabled: PropTypes.bool,
  }

  static defaultProps = {
    onLoadMore: () => {},
    sendQueryParams: () => {},
    children: null,
    activities: [],
    activityTypeFilters: {},
    hasMore: false,
    isLoading: false,
    actions: null,
    totalActivities: 0,
    isGlobalUltimate: false,
    dnbHierarchyCount: null,
    isTypeFilterFlagEnabled: false,
    isGlobalUltimateFlagEnabled: false,
  }

  constructor(props) {
    const {
      allActivity,
      myActivity,
      externalActivity,
      dataHubActivity,
    } = props.activityTypeFilters

    super(props)

    this.state = {
      activityTypeFilter: dataHubActivity ? dataHubActivity.value : '',
      showDnbHierarchy: false,
      showDetails: false,
      activityTypeFilters: [
        allActivity,
        myActivity,
        externalActivity,
        dataHubActivity,
      ],
    }

    this.onActivityTypeFilterChange = this.onActivityTypeFilterChange.bind(this)
    this.onShowDetailsClick = this.onShowDetailsClick.bind(this)
    this.showActivitiesFromAllCompanies = this.showActivitiesFromAllCompanies.bind(
      this
    )
  }

  onActivityTypeFilterChange(e) {
    const activityTypeFilter = e.target.value
    const { sendQueryParams } = this.props
    const { showDnbHierarchy } = this.state

    sendQueryParams({
      activityTypeFilter,
      showDnbHierarchy,
    })

    this.setState({
      activityTypeFilter,
    })
  }

  showActivitiesFromAllCompanies(e) {
    const showDnbHierarchy = e.target.checked
    const { sendQueryParams } = this.props
    const { activityTypeFilter } = this.state

    sendQueryParams({
      activityTypeFilter,
      showDnbHierarchy,
    })

    this.setState({
      showDnbHierarchy,
    })
  }

  onShowDetailsClick(e) {
    this.setState({
      showDetails: e.target.checked,
    })
  }

  render() {
    const {
      activities,
      onLoadMore,
      hasMore,
      isLoading,
      actions,
      children,
      totalActivities,
      isGlobalUltimate,
      dnbHierarchyCount,
      isTypeFilterFlagEnabled,
      isGlobalUltimateFlagEnabled,
    } = this.props

    const {
      activityTypeFilters,
      activityTypeFilter,
      showDetails,
      showDnbHierarchy,
    } = this.state

    const hasFilters = isTypeFilterFlagEnabled || isGlobalUltimateFlagEnabled

    return (
      <ActivityFeedContainer>
        <ActivityFeedHeader
          totalActivities={totalActivities}
          actions={actions}
        />

        {hasFilters && (
          <ActivityFeedFilters
            activityTypeFilters={activityTypeFilters}
            activityTypeFilter={activityTypeFilter}
            onActivityTypeFilterChange={this.onActivityTypeFilterChange}
            showActivitiesFromAllCompanies={this.showActivitiesFromAllCompanies}
            isGlobalUltimate={isGlobalUltimate}
            dnbHierarchyCount={dnbHierarchyCount}
            isTypeFilterFlagEnabled={isTypeFilterFlagEnabled}
            isGlobalUltimateFlagEnabled={isGlobalUltimateFlagEnabled}
          />
        )}

        {activities.length > 0 && (
          <ActivityFeedShowAll
            onChange={this.onShowDetailsClick}
            checked={showDetails}
          >
            Show details for all activities
          </ActivityFeedShowAll>
        )}

        <ActivityFeedCardList>
          {activities.map((activity) => (
            <li key={activity.id}>
              <Activity
                activity={activity}
                showDetails={showDetails}
                showDnbHierarchy={showDnbHierarchy}
              />
            </li>
          ))}
        </ActivityFeedCardList>

        {hasMore && (
          <ActivityFeedPagination
            isLoading={isLoading}
            onLoadMore={onLoadMore}
          />
        )}

        {children}
      </ActivityFeedContainer>
    )
  }
}
