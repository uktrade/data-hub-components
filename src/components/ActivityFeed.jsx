import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { SPACING } from '@govuk-react/constants'

import ActivityFeedCard from './ActivityFeedCard'
import ActivityFeedHeader from './ActivityFeedHeader'
import ActivityFeedPagination from './ActivityFeedPagination'

const ActivityFeedContainer = styled('div')`
  margin: ${SPACING.SCALE_2};
`

const ActivityFeedCardList = styled('ol')`
  list-style-type: none;
  padding: 0;
  
  & > li {
    margin-bottom: ${SPACING.SCALE_2};
  }
`

export default class ActivityFeed extends React.Component {
  static propTypes = {
    activities: PropTypes.arrayOf(PropTypes.object).isRequired,
    onLoadMore: PropTypes.func,
  }

  render() {
    const { activities, onLoadMore, isLoading } = this.props

    return (
      <ActivityFeedContainer>
        <ActivityFeedHeader totalCards={activities.length} />

        <ActivityFeedCardList>
          {activities.map(activity => {
            return <li key={activity.id}>
              <ActivityFeedCard activity={activity} />
            </li>
          })}
        </ActivityFeedCardList>

        {onLoadMore && <ActivityFeedPagination isLoading={isLoading} onLoadMore={onLoadMore}/>}
      </ActivityFeedContainer>
    )
  }
}
