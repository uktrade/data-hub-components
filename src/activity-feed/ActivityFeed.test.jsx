import React from 'react'
import renderer from 'react-test-renderer'
import MockDate from 'mockdate'

import ActivityFeed from './ActivityFeed'
import interactionActivityFixture from '../../fixtures/activity_feed/interactions/interaction'

// Lock the date so moment's relative date doesn't break our deterministic tests.
MockDate.set(1559750582706)

describe('ActivityFeed', () => {
  test('renders empty feed', () => {
    const tree = renderer
      .create(<ActivityFeed />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders single activity', () => {
    const tree = renderer
      .create(<ActivityFeed totalActivities={1} activities={[interactionActivityFixture]}/>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders with "Load more" link', () => {
    const tree = renderer
      .create(<ActivityFeed
        totalActivities={20}
        activities={
          Array.from({ length: 20 }, (e, i) => {
            return {
              ...interactionActivityFixture,
              id: i,
            }
          })
        }
        hasMore={true}
      />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders with loader', () => {
    const tree = renderer
      .create(
        <ActivityFeed
          isLoading={true}
          hasMore={true}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders its children', () => {
    const tree = renderer
      .create(
        <ActivityFeed>
          <div>Child element to test</div>
        </ActivityFeed>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})




