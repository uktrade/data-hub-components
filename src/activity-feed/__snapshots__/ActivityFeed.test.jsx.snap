// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityFeed when there are activities should render activity 1`] = `
<div
  className="sc-daURTG cCbora"
>
  <div
    className="sc-GMQeP cPFqxA"
  >
    <div
      className="sc-exAgwC jnWYar"
    >
      <h2
        className="sc-cMljjf leOqMw"
        size="LARGE"
      >
        1
         activities
      </h2>
    </div>
    <div
      className="sc-cQFLBn cxgntD"
    >
      <button
        className="sc-VigVT kJjasd"
        disabled={false}
      >
        Add an interaction
      </button>
    </div>
  </div>
  <ol
    className="sc-bXGyLb buRwxd"
  >
    <li>
      <div
        className="sc-cmthru jjutsD"
      >
        <div
          className="sc-hMFtBS bSjygG"
        >
          <div
            className="sc-cLQEGU ciAtZf"
          >
            <div
              className="sc-jAaTju fBEFVB"
            >
              <p>
                An interaction with the company took place
              </p>
            </div>
          </div>
          <div
            className="sc-gqPbQI fPqDmG"
          >
            <div
              className="sc-jAaTju fBEFVB"
            >
              <p>
                20 days ago
              </p>
            </div>
          </div>
          <div
            className="sc-hORach eRIqAV"
          >
            <h3
              className="sc-cMljjf kCLKvB"
              size="MEDIUM"
            >
              Invited to Technology Roundtable
            </h3>
          </div>
        </div>
        <details
          className="sc-bMVAic cchcEg sc-kEYyzF ihcnje"
          open={false}
        >
          <summary
            className="sc-kkGfuU cCdZFh"
          >
            <span
              className="sc-iAyFgw jRkIxn"
            >
              Who was involved
            </span>
          </summary>
          <div
            className="sc-hSdWYo dfimje"
          >
            <table
              className="sc-bwCtUz iVExri"
            >
              <tbody
                className="sc-eXEjpC fuVqJC"
              >
                <tr
                  className="sc-iQKALj ftJWrM"
                >
                  <th
                    className="sc-RefOD dxRfSY"
                    style={
                      Object {
                        "border": 0,
                        "fontWeight": "normal",
                      }
                    }
                  >
                    Advisers
                  </th>
                  <td
                    className="sc-RefOD gzbQXK"
                    style={
                      Object {
                        "border": 0,
                      }
                    }
                  >
                    <span>
                      Mario Bros
                       
                      <a
                        className="sc-bwzfXH VXMCV"
                        href="mailto:mariobros@dit.com"
                        muted={false}
                      >
                        mariobros@dit.com
                      </a>
                    </span>
                  </td>
                </tr>
                <tr
                  className="sc-iQKALj ftJWrM"
                >
                  <th
                    className="sc-RefOD dxRfSY"
                    style={
                      Object {
                        "border": 0,
                        "fontWeight": "normal",
                      }
                    }
                  >
                    Services
                  </th>
                  <td
                    className="sc-RefOD gzbQXK"
                    style={
                      Object {
                        "border": 0,
                      }
                    }
                  >
                    Account Management
                  </td>
                </tr>
              </tbody>
            </table>
            <a
              className="sc-bwzfXH VXMCV"
              href="https://example.com/companies/a4ead59f-dfc8-e511-a5e6-e4115bead28a/interactions/a066cc49-b679-4cbe-93c5-3ceba2f067ec"
              muted={false}
            >
              Go to the interaction detail page
            </a>
          </div>
        </details>
      </div>
    </li>
  </ol>
</div>
`;

exports[`ActivityFeed when there are no activities should render empty feed 1`] = `
<div
  className="sc-daURTG cCbora"
>
  <div
    className="sc-GMQeP cPFqxA"
  >
    <div
      className="sc-exAgwC jnWYar"
    >
      <h2
        className="sc-cMljjf leOqMw"
        size="LARGE"
      >
        0
         activities
      </h2>
    </div>
    <div
      className="sc-cQFLBn cxgntD"
    >
      <button
        className="sc-VigVT kJjasd"
        disabled={false}
      >
        Add an interaction
      </button>
    </div>
  </div>
  <ol
    className="sc-bXGyLb buRwxd"
  />
</div>
`;
